### PR TITLE
make k8s the default builder

### LINKED ISSUE #501: Make kubernetes the default builder

### PR DESCRIPTION
**Pull request by mojtaba-esk**
Closes #501 

Currently knuu is using k8s builder (kaniko) by default, however when a change is done to an image and the image has to be rebuild, knuu uses docker locally to rebuild the modified image (i.e. dockerfile), this PR proposes a solution where the new changes also are built using kaniko which is done on k8s instead of local.
Another thing this PR is proposes is when user wants to extract a file from an image, i.e. when an instance is not running, but user wants to read a file from it. That was done locally by docker, this PR also suggests a solution for it to be done on k8s.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new test function to validate file downloading from a builder instance.
	- Added a method for retrieving files directly from container images, enhancing file access capabilities.

- **Bug Fixes**
	- Removed retry logic in a test function for setting the Git repository, simplifying control flow but potentially affecting test reliability in transient error scenarios.

- **Refactor**
	- Simplified command execution process and Docker image building workflow, focusing on Dockerfile instruction management without direct Docker client interactions.
	- Updated methods to minimize error handling, reflecting a shift towards a less defensive programming style.

### CODE REVIEW

File: pkg/container/docker.go [hunk 1]
```diff
@@ -55,104 +42,24 @@ func (f *BuilderFactory) ImageNameFrom() string {
 	return f.imageNameFrom
 }
 
-// ExecuteCmdInBuilder runs the provided command in the context of the given builder.
-// It returns the command's output or any error encountered.
-func (f *BuilderFactory) ExecuteCmdInBuilder(command []string) (string, error) {
+// AddCmdToBuilder runs the provided command in the context of the given builder.
```
Comment (MSevey) [inline pkg/container/docker.go:45]:
```suggestion
// AddCmdToBuilder adds the provided command to be run in the context of the builder.
```

**Review by MSevey — COMMENTED** [review]
over lgtm, do we have an image rebuild test that would cover these changes?

**Comment by mojtaba-esk** [comment]
> over lgtm, do we have an image rebuild test that would cover these changes?

There are already a couple of tests that builds images and they are passing using the new change. Knuu has been using kaniko builder since long time ago, this PR proposes the changes to the underlying layers where still docker used to rebuild an image which is modified. e.g. a file added and there is a new test to test that:

```go
func (s *Suite) TestDownloadFileFromBuilder()
```

**Review by MSevey — DISMISSED** [review]

File: pkg/instance/build.go [hunk 2]
```diff
@@ -133,9 +130,7 @@
 		return ErrSettingUserNotAllowed.WithParams(b.instance.state.String())
 	}
 
-	if err := b.builderFactory.SetUser(user); err != nil {
-		return ErrSettingUser.WithParams(user, b.instance.name).Wrap(err)
-	}
+	b.builderFactory.SetUser(user)
```
[the remark this answers is not in the document]
Comment (mojtaba-esk) [reply pkg/instance/build.go:133]: with the recent merge, `b.builderFactory.SetUser(user)` does not return anything.

File: pkg/instance/build.go [hunk 3]
```diff
@@ -224,14 +219,10 @@
...
-	_ = src
+func (b *build) addFileToBuilder(src, dest, chown string) {
 	// dest is the same as src here, as we copy the file to the build dir with the subfolder structure of dest
-	err := b.builderFactory.AddToBuilder(dest, dest, chown)
-	if err != nil {
-		return ErrAddingFileToInstance.WithParams(dest, b.instance.name).Wrap(err)
-	}
-	return nil
+	_ = src
+	b.builderFactory.AddToBuilder(dest, dest, chown)
```
[the remark this answers is not in the document]
Comment (mojtaba-esk) [reply pkg/instance/build.go:225]: same here. it does not return an error.

File: pkg/instance/build.go [hunk 4]
```diff
@@ -247,7 +238,8 @@
 	}).Debugf("Setting environment variable")
 
 	if b.instance.state == StatePreparing {
-		return b.builderFactory.SetEnvVar(key, value)
+		b.builderFactory.SetEnvVar(key, value)
+		return nil
```
[the remark this answers is not in the document]
Comment (mojtaba-esk) [reply pkg/instance/build.go:242]: same here. it does not return an error.

File: pkg/instance/storage.go [hunk 5]
```diff
@@ -47,7 +48,8 @@ func (s *storage) AddFile(src string, dest string, chown string) error {
 
 	switch s.instance.state {
 	case StatePreparing:
-		return s.instance.build.addFileToBuilder(src, dest, chown)
+		s.instance.build.addFileToBuilder(src, dest, chown)
+		return nil
```
[the remark this answers is not in the document]
Comment (mojtaba-esk) [reply pkg/instance/storage.go:52]: same here. it does not return an error.

File: pkg/instance/build.go [hunk 6]
```diff
@@ -119,10 +119,7 @@ func (b *build) ExecuteCommand(command ...string) error {
 		return ErrAddingCommandNotAllowed.WithParams(b.instance.state.String())
 	}
 
-	_, err := b.builderFactory.ExecuteCmdInBuilder(command)
-	if err != nil {
-		return ErrExecutingCommandInInstance.WithParams(command, b.instance.name).Wrap(err)
-	}
+	b.builderFactory.AddCmdToBuilder(command)
```
[the remark this answers is not in the document]
Comment (mojtaba-esk) [reply pkg/instance/build.go:122]: same here. it does not return an error.

**Pull request merged**